hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/blockmanagement/NumberReplicasStriped.java

```diff
@@ -0,0 +1,137 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.hadoop.hdfs.server.blockmanagement;
+
+import org.apache.hadoop.thirdparty.com.google.common.base.Objects;
+
+import java.util.concurrent.ThreadLocalRandom;
+
+import static org.apache.hadoop.hdfs.server.blockmanagement.NumberReplicas.StoredReplicaState.LIVE;
+
+class NumberReplicasStriped extends NumberReplicas {
+
+  private DatanodeStorageInfo[] storages;
+  private StoredReplicaState[] states;
+  private boolean considerBusy;
+  private boolean[] busys;
+
+  NumberReplicasStriped(int totalBlock, boolean considerBusy) {
+    this.storages = new DatanodeStorageInfo[totalBlock];
+    this.states = new StoredReplicaState[totalBlock];
+    this.considerBusy = considerBusy;
+    if (this.considerBusy) {
+      this.busys = new boolean[totalBlock];
+    }
+  }
+
+  public void add(final StoredReplicaState state, final long value, int blockIndex,
+      DatanodeStorageInfo storage, boolean busy) {
+    if (this.storages[blockIndex] == null) {
+      this.storages[blockIndex] = storage;
+      this.states[blockIndex] = state;
+      if (this.considerBusy) {
+        this.busys[blockIndex] = busy;
+      }
+      this.add(state, 1);
+    } else if (state.getPriority() < this.states[blockIndex].getPriority()) {
+      this.subtract(this.states[blockIndex], 1);
+      this.storages[blockIndex] = storage;
+      this.states[blockIndex] = state;
+      if (this.considerBusy) {
+        this.busys[blockIndex] = busy;
```

Comment:
busys should not set in NumberReplicasStriped. we need a wrapper. I will change it if the proposal is accepted.